tests/__init__.py

```diff
@@ -76,7 +76,18 @@ def run(cwd, exe, args, env=dict(os.environ), **kwargs):
             coverage_dir,
             *cmd,
         ]
-    return subprocess.run([*cmd, *args], cwd=cwd, env=env, **kwargs)
+    if "valgrind" in os.environ.get("RUN_ANALYZER", ""):
+        cmd = ["valgrind", "--leak-check=yes", *cmd]
+    try:
+        failed = False
+        return subprocess.run([*cmd, *args], cwd=cwd, env=env, **kwargs)
+    except subprocess.CalledProcessError:
```

Comment:
Personally I frown on this because the control flow becomes really hard and it's not really obvious what happens in all cases.  E.g. here the `if failed` statement below is rather abiguous.  I think this is better expressed as
```
try:
    ret = subprocess.run(....)
except subprocess.CalledProcessError:
    ....
    pytest.fail(...)
else:
    return ret
```

Reply:
I should have added a comment about this. This is basically to work around pythons automatic error chaining. The backtrace is horribly unreadable as it is, and this tries to clean it up a bit.